### PR TITLE
Swift 3 migration

### PR DESCRIPTION
**Pull request by ikesyo**
#1558 

__TODO__

- [ ] Pin `ReactiveTask`
- [ ] Pin `PrettyColors`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -36,7 +36,7 @@ There are multiple options for installing Carthage:
 
 * **Homebrew:** You can use [Homebrew](http://brew.sh) and install the `carthage` tool on your system simply by running `brew update` and `brew install carthage`. (note: if you previously installed the binary version of Carthage, you should delete `/Library/Frameworks/CarthageKit.framework`).
 
-* **From source:** If you’d like to run the latest development version (which may be highly unstable or incompatible), simply clone the `master` branch of the repository, then run `make install`. Requires Xcode 8.0 or later (Swift 2.3, but not currently 3.x).
+* **From source:** If you’d like to run the latest development version (which may be highly unstable or incompatible), simply clone the `master` branch of the repository, then run `make install`. Requires Xcode 8.2 (Swift 3.0.2).
```
Comment (mdiep) [inline README.md]: Does this actually require Xcode 8.2/Swift 3.0.2? Or is Xcode 8.0/Swift 3.0 enough?

File: Source/CarthageKit/Swift3Shims.swift [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import Foundation
 
 #if swift(>=3)
```
Comment (mdiep) [inline Source/CarthageKit/Swift3Shims.swift]: We should be able to remove this file?

**Comment by mdiep** [comment]
Ah, looks like there are still a few more things that need to be converted.

File: README.md (hunk 1 above)
Comment (ikesyo) [inline README.md]: Does the fix warrant the requirement?

https://developer.apple.com/library/content/releasenotes/DeveloperTools/RN-Xcode/Introduction.html#//apple_ref/doc/uid/TP40001051-CH1-SW10

> NSFileManager's enumerate(at:includingPropertiesForKeys:options:errorHandler:) method passes valid URLs to the error handler in code compiled with Swift 3.0.2.

File: Source/CarthageKit/Swift3Shims.swift (hunk 2 above)
Comment (ikesyo) [inline Source/CarthageKit/Swift3Shims.swift]: We still need the `Result.fanout` shim (which is not included in 3.1.0) and I wanted to minimize the diff in this pull request. Let's address it in a separate pull request.

**Comment by mdiep** [comment]
Most of the integration tests are failing. 😧

**Comment by ikesyo** [comment]
Weird, those pass locally... 🤔

**Comment by ikesyo** [comment]
Found something odd: https://travis-ci.org/ikesyo/Carthage/builds/196473060#L2744-L2755

**Comment by mdiep** [comment]
I get this when I try to run your branch in Xcode:

```
dyld: Symbol not found: __swift_deallocObject
  Referenced from: /Users/mdiep/Library/Developer/Xcode/DerivedData/Carthage-czpjpropnpyltxevlojqzwlxgsmd/Build/Products/Debug/Runes.framework/Versions/A/Runes
  Expected in: /Library/Frameworks/CarthageKit.framework/Versions/Current/Frameworks/libswiftCore.dylib
 in /Users/mdiep/Library/Developer/Xcode/DerivedData/Carthage-czpjpropnpyltxevlojqzwlxgsmd/Build/Products/Debug/Runes.framework/Versions/A/Runes
```

**Comment by mdiep** [comment]
If I remove my installed Carthage, something seems very upset about [this line](https://github.com/Carthage/Carthage/blob/fcd7a718cc71ead02eb234dba13ca49f80dd137b/Source/carthage/Extensions.swift#L25).

**Comment by mdiep** [comment]
```
BUG IN CLIENT OF LIBDISPATCH: Cannot change the target of this queue after it has been activated
```

**Comment by mdiep** [comment]
Gonna go ahead and merge this so people can update their PRs. No use keeping this open longer than necessary.

We can pin ReactiveTask and PrettyColors in a subsequent PR.

**Comment by ikesyo** [comment]
@mdiep Thanks for the investigation and the fix! 💖

**Pull request merged**